### DISCUSSION TITLE
Multi GPU support

### DISCUSSION THREAD

**Discussion by vs30217** [immich-app/immich#13140] [General]
I am using version 1.117.0 with buffalo_l model for facial recognition use case. My system is having 2 vGPU (Nvidia A40) each having 4G VRAM. I am getting out of memory exception in machine learning container for GPU:0. As my vm is having 2 gpu but model always uses GPU:0. is this is the intended behaviour ? or is it possible to use multiple gpu for facial recognition use case.

**Comment by Mraedis** [comment]
Yes, this is the intended behavior, some ask for multi-GPU support for faster ML, but no work has been done or planned.

Though even with multi-GPU support I'm not sure 4+4 = 8  for the facial grouping usecase?

**Reply by vs30217** [reply]
Thank you for the clarification.

**Comment by bo0tzz** [comment]
If you want to use multiple GPUs, you'll need to spin up an ML container for each and a loadbalancer to spread requests between them.

**Reply by vs30217** [reply]
can you please guide me more on that ?

**Reply by bo0tzz** [reply]
Not really, since it's an advanced thing specific to your setup. If you have a search through the github discussions there's a few people who have done something similar for a remote ML container.

**Comment by bo0tzz** [comment]
#13237

**Comment by rhamblen** [comment]
**Supporting the "don't implement multi-GPU" position — but the docs currently contradict it**

I want to add a data point that supports leaving multi-GPU unimplemented, while flagging that the documentation actively leads users into attempting it.

The [[hardware-acceleration docs](https://docs.immich.app/features/ml-hardware-acceleration/)](https://docs.immich.app/features/ml-hardware-acceleration/) currently say:

> Set `MACHINE_LEARNING_DEVICE_IDS` to a comma-separated list of device IDs and set `MACHINE_LEARNING_WORKERS` to the number of listed devices. The machine learning service will spawn workers, one for each device… **different requests will be processed by one worker or the other.**

> You should **increase job concurrencies** to increase overall utilization and more effectively distribute work across multiple GPUs.

Read alongside the maintainer statement in this thread that single-GPU is intended and unplanned, those two things can't both be true. I followed the docs exactly and can confirm the docs are the inaccurate half.

**Setup:** Immich v3.1.0, `ghcr.io/imagegenius/immich:cuda`, 2× 24 GB GPUs (RTX PRO 4000 Blackwell + RTX 3090), `NVIDIA_VISIBLE_DEVICES` listing both UUIDs, `MACHINE_LEARNING_DEVICE_IDS=0,1`, `MACHINE_LEARNING_WORKERS=2`.

**The configuration works as documented, up to a point.** Two ML worker processes spawn, and each loads its models onto a *different* card — confirmed by mapping PIDs to devices:

```
GPU 0 (Blackwell)  worker pid 316738  3,940 MiB
GPU 1 (RTX 3090)   worker pid 316737  2,724 MiB
```

**But requests never reach worker 1.** Smart-search backlog of ~154k assets, model `ViT-SO400M-16-SigLIP2-384__webli`:

| smartSearch concurrency | throughput | GPU 0 | GPU 1 |
|---|---|---|---|
| 6 | 545/min | 100% | 4–8% |
| 12 | 546/min | 100% | 4–8% |

Doubling concurrency — the docs' specific remedy — produced **no throughput change whatsoever**. The only observable effect was GPU 0's VRAM climbing 18.3 → 22.4 GB of 24.5 GB, i.e. approaching OOM for zero benefit. I reverted it.

My guess is that gunicorn balances per *connection* rather than per *request*, so a keep-alive socket from the server pins every request to whichever worker accepted it — but I haven't confirmed that, and it doesn't really matter for the outcome.

**Suggestion:** rather than implementing multi-GPU, amend the docs to match the intended behaviour:

- State plainly that ML load is **not** distributed across GPUs, and that this is by design
- Note that `DEVICE_IDS` + `WORKERS` will spawn per-device workers and allocate models to each, but requests will not be spread between them — so the extra VRAM is consumed for no gain
- Drop or qualify the "increase job concurrencies to distribute work across multiple GPUs" advice, since it doesn't distribute and can push a single GPU toward OOM
- Point users wanting multi-GPU at the separate-ML-container-plus-load-balancer approach already suggested in this thread

That's a docs-only change and would have saved me a fair amount of time chasing a configuration that was already correct.
